### DISCUSSION TITLE
Installation on Freepascal

### DISCUSSION THREAD

**Discussion by SirTwistDo** [graphics32/graphics32#333] [General] [answered]
Good evening, or Good morning, or whatsoever,

I was just trying to install graphics32 on Freepascal/Windows, and encountered a couple of errors.
First, the installation documentation says to add only the design time package, but compiling this failed, because it cannot find the GR32_Lazarus:
`GR32.Design.Reg.pas(54,3) Error: Kann GR32 nicht finden verwendet von GR32.Design.Reg. Prüfen, ob Package GR32_Lazarus die GR32.ppu erzeugt, nichts prüfen löscht diese Datei und prüft, daß keine zwei Packages Zugriff auf die Unit-Quellen haben..`

So I tried to compile the package GR32_Lazarus, and again I got a couple of errors. A function in TLayerCollection
` function  ViewportToLocal(const APoint: TFloatPoint; AScaled: Boolean): TFloatPoint;`
expects a TFloatPoint, but is called somewhere with a TPoint. There is no implicit convertion implemented. I solved it with an overloaded version of ViewportToLocal. 

Then, TLayerCollection seems to have some generic methods defined, like Add and Insert, but Freepascal does not support this kind of syntax. I removed them temporarly, resp. commented it out

Then, in GR32.transpose, in lines 187-189, I got
```GR32.Transpose.pas(187,3) Error: Asm: [push reg??] invalid combination of opcode and operands
GR32.Transpose.pas(188,3) Error: Asm: [push reg??] invalid combination of opcode and operands
GR32.Transpose.pas(189,3) Error: Asm: [push reg??] invalid combination of opcode and operands
GR32.Transpose.pas(526,3) Error: Asm: [pop reg??] invalid combination of opcode and operands
GR32.Transpose.pas(527,3) Error: Asm: [pop reg??] invalid combination of opcode and operands
GR32.Transpose.pas(528,3) Error: Asm: [pop reg??] invalid combination of opcode and operands
```

And here I have to admit, I cannot continue.

Any help is highly appreciated! I dont know if this is a local problem, and others are able to install graphics32?
FPC 3.2.2, Lazarus 3.0. Windows 10.

Kind regards,
Michael

**Comment by andersmelander** [comment]
Yes, I'm afraid the FPC support has fallen somewhat behind and I know it's broken at the moment.

One of the problems is that I'm not really using FPC for anything other than keeping Graphics32 FPC compatible and various deficiencies in the FPC compiler (see #297 for example) makes this very hard.

I will see if I can fix it in a few days. FWIW, I'm using Lazarus 3.2 with FPC 3.2.2 on  Windows 10.

**Reply by SirTwistDo** [reply]
Good Morning, thank you for that fast reply! Is there anything I can support you with? I am using FPC/Lazarus on a daily base, but for a new (private) project I reached the limits of the standard TBitmap/Canvas, and someone pointed me to your Graphics32 Library.

**Reply by andersmelander** [reply]
Good morning,

Thanks for the offer of assistance but I think that the majority of work will be to get the assembler to compile. Luckily I have provided pure pascal alternatives for all the assembler functions so if all else fails I can just disable the assembler.
I'll have something for you to try tonight or tomorrow.

**Comment by andersmelander** [comment, marked as answer]
Okay, so the [FPC branch](https://github.com/graphics32/graphics32/tree/FPC) should compile with Lazarus 3.2. It probably also works with Lazarus 3.0 - give it a try.

I have **not** tested these changes very thoroughly but if you find something that doesn't work just let me know and I'll look at it.

The design-time installation procedure for Lazarus is pretty funky due to the FPC inline bug but this works for me most of the time:

1. Iteration 1
    a. Open the Source\Packages\Lazarus\GR32_Lazarus package
    b. Compile clean
    c. Open the Source\Packages\Lazarus\GR32_DSGN_Lazarus package
    d. Use/Install, yes to rebuild Lazarus
    e. Compile **fails** with inline bug

2. Iteration 2
    a. Compile (not clean) GR32_Lazarus package
    b. Use/Install GR32_DSGN_Lazarus package, yes to rebuild Lazarus
    c. Compile **fails** with inline bug

3. Iteration 3
    a. Compile (not clean) GR32_Lazarus package
    b. Use/Install GR32_DSGN_Lazarus package, yes to rebuild Lazarus
    c. Compile **succeeds**. Lazarus relinks and restarts.

**Reply by SirTwistDo** [reply]
Hey Anders, unfortunately it is not working with me. I removed my complete FPC/Lazarus installation and installed it freshly from lazarus-3.2-fpc-3.2.2-win64.exe, but I am still getting the ASM errors. I added then the "PUREPASCAL" define, now I get:
`GR32.Transpose.pas(704,18) Error: Identifier not found "TInterlocked"`
I found this https://gitlab.com/freepascal.org/fpc/source/-/issues/37089 and it says "TInterlocked" is implemented in FPC 3.3.1
I am stuck installing FPC3.3.1 + Laz 3.2 or Laz 3.4 (using fpcupdeluxe).

**Reply by andersmelander** [reply]
You need to use the FPC branch; I haven't merged the changes I made into the master branch yet.

In the FPC branch the `TInterlocked.Exchange` has been conditionally replaced with `InterlockedExchange` for FPC:
https://github.com/graphics32/graphics32/blob/c51d753745d106cf3958ec37588d8b5f4e081830/Source/GR32.Transpose.pas#L706-L710

I just pushed a batch of new commits to the FPC branch that resolves the remaining, known FPC assembler issues. These were a b*tch to fix as the Lazarus debugger really sucks with SSE2 code (or just assembler in general). I updated my Lazarus to 3.6 in the hope that it had improved since 3.2 - but it hadn't :-/

**Reply by SirTwistDo** [reply]
Good Morning, I am sorry, I haven't noticed that you mentioned the FPC branch :-(
With Laz 3.0, nearly everything worked fine, except that I had to include "GR32.Blend.Assembler.pas" to GR32 to compile GR32_Design. GR32 compiles without that Assembler file, but the design package not. Regardless wether I set "PUREPASCAL" or not.
Thank you very much for your quick support!!! Next weekend with bad weather, I will also upgrade to Laz 3.4 or 3.6.

**Reply by andersmelander** [reply]
@SirTwistDo FYI, the FPC branch is waiting for your test; I won't merge it into master until it has been tested by someone actually using FPC.

**Reply by SirTwistDo** [reply]
> @SirTwistDo FYI, the FPC branch is waiting for your test; I won't merge it into master until it has been tested by someone actually using FPC.

Hello Anders, thank you for the reminder. As I wrote, I needed to include "GR32.Blend.Assembler.pas" to GR32. 
I played around a lot with TBitmap32 and some renderes and didn't encountered any problems. I haven't upgraded Lazarus yet. Some of the examples are not compiling, but those which I could compile were working fine.

**Reply by andersmelander** [reply]
That's good to hear.

The GR32_Lazarus package already contains `GR32.Blend.Assembler` so if you had to manually add that then you weren't using the latest version of the [FPC branch](https://github.com/graphics32/graphics32/tree/FPC):
https://github.com/graphics32/graphics32/blob/b89ac8b2beea6f31e798131c5a8055eb19cf66b2/Source/Packages/Lazarus/GR32_Lazarus.pas#L20

I have just gone through all the 41 examples which have Lazarus projects files and they should compile and run without problems. Most of the rest should too but I just haven't bothered creating Lazarus project files for them. Often it just involves importing the Delphi project into Lazarus and deleting the lfm file. Now that Lazarus can read the Delphi dfm files directly it's gotten a lot easier to keep the examples compatible.

Anyway, there are a few minor things I'd like to add for FPC and after that I'll merge - in a couple of days.

**Reply by SirTwistDo** [reply]
Good Morning. I've just got the latest FPC branch. Compilation of GR32 worked well, but the design package wouldn't compile (as part of a Lazarus rebuild). I had to add "..\.." to the Compilersettings->Paths, to the first field "-Fu".
Regarding the examples:
* Examples\Drawing\Curves: Lazarus complains about invalid attributes in Main.lfm, like AlignWithMargins, Margins, TextHeight, ShowCaption
* Examples\Drawing\MeshGradients: Unit output directory "..\..\..\Unit Cache\$(TargetCPU)-$(TargetOS)" is missing. Usual value is "lib\$(TargetCPU)-$(TargetOS)"
* Examples\General\Image32 background: 
```
Verbose: Compiling resource P:\Develop\lib\Graphics32\Examples\General\Image32 background\Media.rc
Warning: windres [option(s)] [input-file] [output-file]
```
* Examples\Resampling\NestedSampling: again missing unit output directory with that "Unit Cache". I gues the space is the problem here. I replaced it with the usual "lib....." and it compiled

Once again, thank you very much for your support and for that GR32 library at all!!

**Reply by andersmelander** [reply]
- > I had to add "..\\.." to the Compilersettings->Paths, to the first field "-Fu".

  Reproduced and fixed.

- The remaining `...\Unit Cache\...`references should now have been replaced with `lib\...`

- > Lazarus complains about invalid attributes in *.dfm

  These are harmless and there's not much I can do about it since there will always be differences in what Delphi and Lazarus put in the form files. I could go back to having separate dfm and lfm files but then the lfm files are bound to get out of sync again.

- >Warning: windres [option(s)] [input-file] [output-file]

  I'm not seeing that warning. The `lib\media.res` file is produced as expected.

Thanks for the feedback

**Comment by andersmelander** [comment]
I think I've managed to get rid of the dependencies that triggered the Lazarus inlining bug; On my system the run- and design-time packages now build and link in one go, without any work-around.
